plugins/org.eclipse.xtext.util/src/org/eclipse/xtext/util/OnChangeEvictingCache.java

```diff
@@ -109,11 +110,19 @@ protected void cacheHit(CacheAdapter adapter) {
 	 * @return the cache adapter for the given resource. Never <code>null</code>.
 	 */
 	public CacheAdapter getOrCreate(Resource resource) {
-		CacheAdapter adapter = (CacheAdapter) EcoreUtil.getAdapter(resource.eAdapters(), CacheAdapter.class);
-		if (adapter == null) {
-			adapter = new CacheAdapter();
-			resource.eAdapters().add(adapter);
-			adapter.setResource(resource);
+		// ask for the list of adapters twice in order to avoid the usage of a
+		//  potentially stale list because of a race of several involved threads
+		resource.eAdapters();
```

Comment:
I'm absolutely aware of that scenario, and I discussed that problem with Moritz. Nonetheless, the list is most likely already available because of the uri change adapter installed by the XtextResourceSet. Can the access to the adapters list of arbitrary resources be coordinated, at all?